### PR TITLE
feat(ouroborus): container characterization

### LINKED ISSUE #25: Object that describe desired container

### PR DESCRIPTION
**Pull request by jeffque**
closes #25

### CODE REVIEW

File: ouroborus/lib/ouroborus/container.rb [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      return "''" if arg.empty?
+      arg.chars.reduce '' do | acc, curr |
+        next "#{acc}\\#{curr}" if backslasheable? curr
+        "#{acc}#{curr}"
+      end
+    end
+
+    def backslasheable? c
+      ' \\$#\'\"'.include? c
+    end
```
Comment (jeffque) [inline ouroborus/lib/ouroborus/container.rb:44]: Implementation detail, shall not be exposed. Private

File: ouroborus/test/test_container.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+require 'rspec/autorun'
+
+describe Ouroborus::Args do
+  it 'should escape quotes, space, dolar, hash, bang and brackets' do
+    args = Ouroborus::Args.new
+    args << 'abc'
+    args << 'lalala lelele'
+    expect(args.to_s).to eq('abc lalala\\ lelele')
+  end
+end
```
[the remark this answers is not in the document]
Comment (jeffque) [reply ouroborus/test/test_container.rb:14]: It gets easier to ready with an empty line between those...

I hava doubt if it is more idiomatic to add a new test file or to keep it all together here, as both groups are about `ouroborus/container.rb`

File: ouroborus/lib/ouroborus/container.rb [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    def port(incoming, interface = nil)
+      p = "#{incoming}"
+      p += ":#{interface}" unless interface.nil?
+      @dockerArgs << '-p' << p
+    end
+
+    def volume(their, ours = nil)
+      v = "#{their}"
```
Comment (jeffque) [inline ouroborus/lib/ouroborus/container.rb:113]: Check if paths are absolute?

File: ouroborus/lib/ouroborus/container.rb [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        "#{RestartCondition.as_string condition}:#{maxRetries}"
+      else
+        "#{RestartCondition.as_string condition}"
+      end
+      @dockerArgs << '--restart' << condStr
+    end
+
+    def env(name, value = nil)
+      @dockerArgs << '--env' << if name.is_a? Hash
+        raise "When first arg is a hash, env second arg should be nil" unless value.nil?
```
Comment (EronAlves1996) [inline ouroborus/lib/ouroborus/container.rb:131]: I think that a message like "When first arg passed is a hash, env second arg should not be passed" can be better. 

What do you think about that?

File: ouroborus/lib/ouroborus/container.rb [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            envvalue
+          end
+        end
+      elsif name.is_a? Array
+        usedValue = if value.nil? then
+          []
+        elsif value.is_a? Array then
+          usedValue = value
+        else
+          raise "When first arg is an Array, env second arg must be another array or nil"
```
Comment (EronAlves1996) [inline ouroborus/lib/ouroborus/container.rb:147]: Just like above: raise "When first arg is an Array, env second arg must be another array or __should not be passed__"

File: ouroborus/lib/ouroborus/container.rb (hunk 4 above)
Comment (jeffque) [reply ouroborus/lib/ouroborus/container.rb:131]: Way better, ty

Comment (jeffque) [reply ouroborus/lib/ouroborus/container.rb:131]:
```suggestion
        raise "When first arg passed is a Hash, env second arg should not be passed" unless value.nil?
```

File: ouroborus/lib/ouroborus/container.rb (hunk 5 above)
Comment (jeffque) [reply ouroborus/lib/ouroborus/container.rb:147]:
```suggestion
          raise "When first arg is an Array, env second arg must be another array or should not be passed"
```

**Pull request merged**